backend/WebApi/Validation/RequireProjectCodeFilter.cs

```diff
@@ -0,0 +1,44 @@
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.Filters;
+using Microsoft.IdentityModel.Tokens;
+using static LanguageForge.WebApi.Constants;
+
+namespace LanguageForge.WebApi.Validation;
+
+public class RequireProjectCodeFilter : IResourceFilter
+{
+    private readonly ILfProjectContext _lfProjectContext;
+
+    public RequireProjectCodeFilter(ILfProjectContext lfProjectContext)
+    {
+        _lfProjectContext = lfProjectContext;
+    }
+
+    public void OnResourceExecuted(ResourceExecutedContext context)
+    {
+    }
+
+    public void OnResourceExecuting(ResourceExecutingContext context)
+    {
+        if (_lfProjectContext.ProjectCode.IsNullOrEmpty())
+        {
+            var requiresProjectCode = context.ActionDescriptor.FilterDescriptors.Any(x => x.Filter is RequireProjectCodeAttribute);
+            if (requiresProjectCode)
+            {
+                context.Result = new BadRequestObjectResult($"{ProjectCodeHeader} header is required");
+            }
+        }
+    }
+}
+
+[AttributeUsage(AttributeTargets.Class | AttributeTargets.Method, Inherited = true)]
+public class RequireProjectCodeAttribute : Attribute, IResourceFilter
```

Comment:
We can't inject dependencies into a custom attribute-filter. So we either:

1) Create a marker attribute filter (which I did, but is probably a bit silly and not the best choice performance wise)
2) Do this: `[ServiceFilter(typeof(RequireProjectCodeFilter))]`, which is perfectly fine, just a bit ugly 🤷 
3) Or pull the `ProjectContext` out of the service provider at run-time.

I guess if the only complaint against 2 is that it's ugly then...we should probably go with that?

Reply:
so this would be instead of `RequireProjectCodeFilter` right?
I think having a dependency on `ILfProjectContext` is not required. Just do a simple check if the header exists and has a value. Like this:
```C#
if (!context.HttpContext.Request.Headers.TryGetValue(ProjectCodeHeader, out var projectCodeHeader) || string.IsNullOrEmpty(projectCodeHeader))
{
    context.Result = new BadRequestObjectResult($"{ProjectCodeHeader} header is required");
}
```

we can put that directly inside the attribute and call it done. If we don't like that there's 2 different places in code (here and in the project context) that retrieve the header value then we can put that in some static helper class and reference it in both places.

Reply:
Yeah that sounds good 👍🏽